core/src/main/java/org/apache/gravitino/policy/PolicyManager.java

```diff
@@ -241,13 +302,112 @@ public String[] associatePoliciesForMetadataObject(
-    throw new UnsupportedOperationException("Not implemented yet");
+    Preconditions.checkArgument(
+        Policy.SUPPORTS_ALL_OBJECT_TYPES.contains(metadataObject.type()),
+        "Cannot associate policies for unsupported metadata object type %s",
+        metadataObject.type());
+
+    NameIdentifier entityIdent = MetadataObjectUtil.toEntityIdent(metalake, metadataObject);
+    Entity.EntityType entityType = MetadataObjectUtil.toEntityType(metadataObject);
+
+    MetadataObjectUtil.checkMetadataObject(metalake, metadataObject);
+
+    // Remove all the policies that are both set to add and remove
+    Set<String> policiesToAddSet =
+        policiesToAdd == null ? Sets.newHashSet() : Sets.newHashSet(policiesToAdd);
+    Set<String> policiesToRemoveSet =
+        policiesToRemove == null ? Sets.newHashSet() : Sets.newHashSet(policiesToRemove);
+    Set<String> common = Sets.intersection(policiesToAddSet, policiesToRemoveSet).immutableCopy();
+    policiesToAddSet.removeAll(common);
+    policiesToRemoveSet.removeAll(common);
+
+    NameIdentifier[] policiesToAddIdent =
+        policiesToAddSet.stream()
+            .map(p -> NameIdentifierUtil.ofPolicy(metalake, p))
+            .toArray(NameIdentifier[]::new);
+    NameIdentifier[] policiesToRemoveIdent =
+        policiesToRemoveSet.stream()
+            .map(p -> NameIdentifierUtil.ofPolicy(metalake, p))
+            .toArray(NameIdentifier[]::new);
+
+    return TreeLockUtils.doWithTreeLock(
+        entityIdent,
+        LockType.READ,
+        () ->
+            TreeLockUtils.doWithTreeLock(
+                NameIdentifier.of(NamespaceUtil.ofPolicy(metalake).levels()),
+                LockType.WRITE,
+                () -> {
+                  try {
+                    checkMetalake(NameIdentifier.of(metalake), entityStore);
+
+                    List<PolicyEntity> updatedPolicies =
+                        entityStore
+                            .relationOperations()
+                            .updateEntityRelations(
+                                SupportsRelationOperations.Type.POLICY_METADATA_OBJECT_REL,
+                                entityIdent,
+                                entityType,
+                                policiesToAddIdent,
+                                policiesToRemoveIdent);
+                    return updatedPolicies.stream().map(PolicyEntity::name).toArray(String[]::new);
+                  } catch (NoSuchEntityException e) {
+                    throw new NoSuchMetadataObjectException(
+                        e,
+                        "Failed to associate policies for metadata object %s due to not found",
+                        metadataObject);
+                  } catch (EntityAlreadyExistsException e) {
+                    throw new PolicyAlreadyAssociatedException(
+                        e,
+                        "Failed to associate policies for metadata object due to some policies %s already "
```

Comment:
What if the policy to be removed does not exist?

Reply:
Do nothing, just the same as the tag operation.